src/passes/JSPI.cpp

```diff
@@ -0,0 +1,169 @@
+
+  JSPI() {}
+
+  Type externref = Type(HeapType::ext, Nullable);
+
+  void run(PassRunner* runner, Module* module) override {
+    Builder builder(*module);
+    // Create a global to store the suspender that is passed into exported
+    // functions and will then need to be passed out to the imported functions.
+    module->addGlobal(builder.makeGlobal(Name("suspender"),
+                                         externref,
+                                         builder.makeRefNull(externref),
+                                         Builder::Mutable));
+
+    // Wrap each exported function in a function that stores the suspender
+    // and calls the original export.
+    std::vector<std::unique_ptr<Export>> newExports;
+    for (auto& ex : module->exports) {
+      if (ex->kind == ExternalKind::Function) {
+        auto* func = module->getFunction(ex->value);
+        auto legalName = makeWrapperForExport(func, module);
+        ex->value = legalName;
+      }
+    }
+    for (auto& ex : newExports) {
+      module->addExport(std::move(ex));
+    }
+
+    // Avoid iterator invalidation later.
+    std::vector<Function*> originalFunctions;
+    for (auto& func : module->functions) {
+      originalFunctions.push_back(func.get());
+    }
+    // Wrap each imported function in a function that gets the global suspender
+    // and passes it on to the imported function.
+    for (auto* im : originalFunctions) {
+      if (im->imported()) {
+        auto funcName = makeWrapperForImport(im, module);
+        ElementUtils::iterAllElementFunctionNames(module, [&](Name& name) {
+          if (name == im->name) {
+            name = funcName;
+          }
+        });
+      }
+    }
+  }
+
+private:
+  Name makeWrapperForExport(Function* func, Module* module) {
+    Name wrapperName(std::string("export$") + func->name.str);
+
+    // a method may be exported multiple times
+    if (module->getFunctionOrNull(wrapperName)) {
+      return wrapperName;
+    }
+
+    Builder builder(*module);
+    auto* wrapperFunc = new Function();
+    wrapperFunc->name = wrapperName;
+
+    auto* call = module->allocator.alloc<Call>();
+    call->target = func->name;
+    call->type = func->getResults();
+
+    // Add an externref param as the first argument and copy all the original
+    // params to new export.
+    std::vector<Type> wrapperParams;
+    wrapperParams.push_back(externref);
+    Index i = 1;
+    for (const auto& param : func->getParams()) {
+      call->operands.push_back(
+        builder.makeLocalGet(wrapperParams.size(), param));
+      wrapperParams.push_back(param);
+      i++;
+    }
+    auto* block = builder.makeBlock();
+    block->list.push_back(
+      builder.makeGlobalSet("suspender", builder.makeLocalGet(0, externref)));
+    block->list.push_back(call);
+    Type resultsType = func->getResults();
+    if (resultsType == Type::none) {
+      // A void return is not currently allowed by v8. Add an i32 return value
```

Comment:
I've filed https://bugs.chromium.org/p/v8/issues/detail?id=13231 about this.

Reply:
Can you add a link to that bug as a code comment? That way we'll be able to easily check whether it has been resolved yet in the future.